sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -560,34 +597,36 @@ public CompletionStage<MutateRowResponse> writeRecord(KV<ByteString, Iterable<Mu
       baseLabels.put(
           MonitoringInfoConstants.Labels.TABLE_ID,
           GcpResourceIdentifiers.bigtableTableID(projectId, instanceId, tableId));
-      ServiceCallMetric serviceCallMetric =
-          new ServiceCallMetric(MonitoringInfoConstants.Urns.API_REQUEST_COUNT, baseLabels);
+      return new ServiceCallMetric(MonitoringInfoConstants.Urns.API_REQUEST_COUNT, baseLabels);
+    }
 
-      CompletableFuture<MutateRowResponse> result = new CompletableFuture<>();
+    private static class WriteMutationCallback implements FutureCallback<MutateRowResponse> {
+      private final CompletableFuture<MutateRowResponse> result;
 
-      outstandingMutations += 1;
-      Futures.addCallback(
-          new VendoredListenableFutureAdapter<>(bulkMutation.add(entry)),
-          new FutureCallback<MutateRowResponse>() {
-            @Override
-            public void onSuccess(MutateRowResponse mutateRowResponse) {
-              result.complete(mutateRowResponse);
-              serviceCallMetric.call("ok");
-            }
-
-            @Override
-            public void onFailure(Throwable throwable) {
-              if (throwable instanceof StatusRuntimeException) {
-                serviceCallMetric.call(
-                    ((StatusRuntimeException) throwable).getStatus().getCode().value());
-              } else {
-                serviceCallMetric.call("unknown");
-              }
-              result.completeExceptionally(throwable);
-            }
-          },
-          directExecutor());
-      return result;
+      private final ServiceCallMetric serviceCallMetric;
+
+      public WriteMutationCallback(
+          CompletableFuture<MutateRowResponse> result, ServiceCallMetric serviceCallMetric) {
+        this.result = result;
+        this.serviceCallMetric = serviceCallMetric;
+      }
+
+      @Override
+      public void onSuccess(MutateRowResponse mutateRowResponse) {
+        result.complete(mutateRowResponse);
+        serviceCallMetric.call("ok");
+      }
+
+      @Override
+      public void onFailure(Throwable throwable) {
+        if (throwable instanceof StatusRuntimeException) {
```

Comment:
I dont think StatusRuntimeExceptions ever leak from the client. I think you want to either look for ApiExceptions or walk the cause chain to find a StatusRuntimeException

Reply:
This is a small refactor of existing code, I didn't actually write this